stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/plugin/factory/StreamOfflinePluginFactory.kt

```diff
@@ -0,0 +1,68 @@
+import io.getstream.chat.android.client.experimental.persistence.OfflinePlugin
+import io.getstream.chat.android.client.experimental.persistence.OfflinePluginFactory
+import io.getstream.chat.android.client.experimental.plugin.factory.PluginFactory
+import io.getstream.chat.android.core.ExperimentalStreamChatApi
+import io.getstream.chat.android.livedata.ChatDomain
+import io.getstream.chat.android.livedata.utils.toLiveDataRetryPolicy
+import io.getstream.chat.android.offline.ChatDomainImpl
+import io.getstream.chat.android.offline.experimental.global.GlobalMutableState
+import io.getstream.chat.android.offline.experimental.plugin.OfflinePluginImpl
+import io.getstream.chat.android.offline.experimental.plugin.configuration.Config
+import io.getstream.chat.android.offline.experimental.plugin.listener.ChannelMarkReadListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.EditMessageListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.GetMessageListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.HideChannelListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.MarkAllReadListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.QueryChannelListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.QueryChannelsListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.listener.ThreadQueryListenerImpl
+import io.getstream.chat.android.offline.experimental.plugin.logic.LogicRegistry
+import io.getstream.chat.android.offline.experimental.plugin.state.StateRegistry
+
+@ExperimentalStreamChatApi
+public class StreamOfflinePluginFactory(
+    private val config: Config,
+    private val appContext: Context,
+) : OfflinePluginFactory, PluginFactory {
+
+    private var instance: OfflinePlugin? = null
+
+    override fun getOrCreate(): OfflinePlugin {
+        return instance ?: createOfflinePlugin().also {
+            instance = it
+        }
+    }
+
+    private fun createOfflinePlugin(): OfflinePlugin {
+        // This can only be called after ChatClient was instantiated!!
+        val chatClient = ChatClient.instance()
+        ChatDomain.Builder(appContext, chatClient).apply {
+            if (config.backgroundSyncEnabled) enableBackgroundSync() else disableBackgroundSync()
+            if (config.persistenceEnabled) offlineEnabled() else offlineDisabled()
+            if (config.userPresence) userPresenceEnabled() else userPresenceDisabled()
+            recoveryEnabled()
+            retryPolicy(config.retryPolicy.toLiveDataRetryPolicy())
+        }.build()
+
+        val stateRegistry =
+            StateRegistry.getOrCreate(io.getstream.chat.android.offline.ChatDomain.instance as ChatDomainImpl)
+        val logic = LogicRegistry.getOrCreate(stateRegistry)
+        val globalStateRegistry = GlobalMutableState.getOrCreate()
```

Comment:
Can we move `GlobalState` to `StateRegistry`? 🤔

Reply:
Now that I took a better look at this class, I'm thinking... Do we even need it? It is a class aggregating the state of the whole SDK... But I feel this could be just an interface implemented by ChatClient, which actually have the state of the SDK already.

Reply:
🤔 currently `ChatClient` doesn't know anything about the "state". All the `StateFlows` related to channel or query channels belong to `OfflinePlugin`